### ISSUE TITLE
glusterfs installation unpredictable

### ISSUE THREAD

**Issue by ryane** [CiscoCloud/microservices-infrastructure#868]
Our glusterfs installation is unreliable. In my experience, 1 out of every 4 or 5 cluster builds fails on glusterfs. Sometimes it is possible to recover by running the Ansible provisioning a second time. In other cases, the only way to (easily [1]) recover is to rebuild the cluster.

This is separate from the single-node glusterfs issues. I am seeing this problem often with 3 control nodes and on multiple platforms.

Let's update this issue with the various errors that occur and try and figure out how we can make the glusterfs install rock-solid.

[1] I am sure it is possible to SSH into the nodes and do some manual work to get glusterfs running. I don't know what those steps are at this point

**Comment by ryane** [comment]
I have a 3-node GCE cluster that failed on glusterfs. Multiple ansible runs do not fix the problem. I neglected to capture the initial error on the first ansible run but subsequent runs return the following:

```
TASK: [glusterfs | create container volume] ***********************************
failed: [resching-gce-control-02] => {"changed": true, "cmd": ["/etc/glusterfs/gluster_volume.sh", "container-volumes"], "delta": "0:00:00.037518", "end": "2015-11-25 14:05:18.214899", "rc": 1, "start": "2015-11-25 14:05:18.177381", "warnings": []}
stderr: + NAME=container-volumes
+ grep -q container-volumes
+ gluster volume list
No volumes present in cluster
+ gluster volume create container-volumes replica 3 resching-gce-control-02.node.consul:/mnt/glusterfs/container-volumes resching-gce-control-01.node.consul:/mnt/glusterfs/container-volumes resching-gce-control-03.node.consul:/mnt/glusterfs/container-volumes
volume create: container-volumes: failed: /mnt/glusterfs/container-volumes is already part of a volume
```

**Comment by justinclift** [comment]
That kind of sounds like a race condition, where a dependency isn't in place by the time it's needed. :frowning:

**Comment by justinclift** [comment]
This error message means the directory it's being given has already been added to a volume:

```
volume create: container-volumes: failed: /mnt/glusterfs/container-volumes is already part of a volume
```

If you want to cleanly reuse that directory, you'll need to wipe it's contents before trying to re-add it.  (eg if it's a volume mount point, remake the volume then remount it.  if it's a plain dir, just rm -rf the dir then mkdir it)

It would also be useful to log into the failed node(s), and look at the gluster log files.  They should be in /var/log/gluster* (I can't remember the exact dir, it's been a while since I touched gluster)

**Comment by justinclift** [comment]
Btw, the Gluster-users mailing list is a good place to seek assistance for the problem when you have the logs in hand. :smile:

&nbsp; &nbsp; http://www.gluster.org/mailman/listinfo/gluster-users

**Comment by ryane** [comment]
Just had a glusterfs failure on an openstack cluster:

```
TASK: [glusterfs | mount container volume] ************************************
failed: [resching-lon1-test-control-02] => {"failed": true}
msg: Error mounting /mnt/container-volumes: Mount failed. Please check the log file for more details.

failed: [resching-lon1-test-control-01] => {"failed": true}
msg: Error mounting /mnt/container-volumes: Mount failed. Please check the log file for more details.

failed: [resching-lon1-test-control-03] => {"failed": true}
msg: Error mounting /mnt/container-volumes: Mount failed. Please check the log file for more details.
```

Here are the contents of the `/var/log/glusterfs/mnt-container-volumes.log` log on all three nodes:

```
$ ansible 'role=control' -s -m shell -a 'cat /var/log/glusterfs/mnt-container-volumes.log'
resching-lon1-test-control-03 | success | rc=0 >>
[2015-12-01 02:41:21.424564] I [MSGID: 100030] [glusterfsd.c:2318:main] 0-/usr/sbin/glusterfs: Started running /usr/sbin/glusterfs version 3.7.6 (args: /usr/sbin/glusterfs --volfile-server=fs.glusterfs.service.consul --volfile-id=/container-volumes /mnt/container-volumes)
[2015-12-01 02:41:26.615024] E [MSGID: 101075] [common-utils.c:306:gf_resolve_ip6] 0-resolver: getaddrinfo failed (Name or service not known)
[2015-12-01 02:41:26.615074] E [name.c:247:af_inet_client_get_remote_sockaddr] 0-glusterfs: DNS resolution failed on host fs.glusterfs.service.consul
[2015-12-01 02:41:26.615180] E [glusterfsd-mgmt.c:1818:mgmt_rpc_notify] 0-glusterfsd-mgmt: failed to connect with remote-host: fs.glusterfs.service.consul (Success)
[2015-12-01 02:41:26.615193] I [glusterfsd-mgmt.c:1824:mgmt_rpc_notify] 0-glusterfsd-mgmt: Exhausted all volfile servers
[2015-12-01 02:41:26.615359] W [glusterfsd.c:1236:cleanup_and_exit] (-->/lib64/libgfrpc.so.0(rpc_clnt_notify+0xd0) [0x7f94431b2d50] -->/usr/sbin/glusterfs(+0xf6be) [0x7f94438bd6be] -->/usr/sbin/glusterfs(cleanup_and_exit+0x69) [0x7f94438b76d9] ) 0-: received signum (1), shutting down
[2015-12-01 02:41:26.615407] I [fuse-bridge.c:5683:fini] 0-fuse: Unmounting '/mnt/container-volumes'.
[2015-12-01 02:41:26.623343] I [MSGID: 101190] [event-epoll.c:632:event_dispatch_epoll_worker] 0-epoll: Started thread with index 1

resching-lon1-test-control-02 | success | rc=0 >>
[2015-12-01 02:41:21.427949] I [MSGID: 100030] [glusterfsd.c:2318:main] 0-/usr/sbin/glusterfs: Started running /usr/sbin/glusterfs version 3.7.6 (args: /usr/sbin/glusterfs --volfile-server=fs.glusterfs.service.consul --volfile-id=/container-volumes /mnt/container-volumes)
[2015-12-01 02:41:26.592676] E [MSGID: 101075] [common-utils.c:306:gf_resolve_ip6] 0-resolver: getaddrinfo failed (Name or service not known)
[2015-12-01 02:41:26.592743] E [name.c:247:af_inet_client_get_remote_sockaddr] 0-glusterfs: DNS resolution failed on host fs.glusterfs.service.consul
[2015-12-01 02:41:26.592866] I [MSGID: 101190] [event-epoll.c:632:event_dispatch_epoll_worker] 0-epoll: Started thread with index 1
[2015-12-01 02:41:26.592926] E [glusterfsd-mgmt.c:1818:mgmt_rpc_notify] 0-glusterfsd-mgmt: failed to connect with remote-host: fs.glusterfs.service.consul (Success)
[2015-12-01 02:41:26.592937] I [glusterfsd-mgmt.c:1824:mgmt_rpc_notify] 0-glusterfsd-mgmt: Exhausted all volfile servers
[2015-12-01 02:41:26.593662] W [glusterfsd.c:1236:cleanup_and_exit] (-->/lib64/libgfrpc.so.0(rpc_clnt_notify+0xd0) [0x7fa4a766cd50] -->/usr/sbin/glusterfs(+0xf6be) [0x7fa4a7d776be] -->/usr/sbin/glusterfs(cleanup_and_exit+0x69) [0x7fa4a7d716d9] ) 0-: received signum (1), shutting down
[2015-12-01 02:41:26.593711] I [fuse-bridge.c:5683:fini] 0-fuse: Unmounting '/mnt/container-volumes'.

resching-lon1-test-control-01 | success | rc=0 >>
[2015-12-01 02:41:21.419710] I [MSGID: 100030] [glusterfsd.c:2318:main] 0-/usr/sbin/glusterfs: Started running /usr/sbin/glusterfs version 3.7.6 (args: /usr/sbin/glusterfs --volfile-server=fs.glusterfs.service.consul --volfile-id=/container-volumes /mnt/container-volumes)
[2015-12-01 02:41:26.609427] E [MSGID: 101075] [common-utils.c:306:gf_resolve_ip6] 0-resolver: getaddrinfo failed (Name or service not known)
[2015-12-01 02:41:26.609495] E [name.c:247:af_inet_client_get_remote_sockaddr] 0-glusterfs: DNS resolution failed on host fs.glusterfs.service.consul
[2015-12-01 02:41:26.609760] I [MSGID: 101190] [event-epoll.c:632:event_dispatch_epoll_worker] 0-epoll: Started thread with index 1
[2015-12-01 02:41:26.609766] E [glusterfsd-mgmt.c:1818:mgmt_rpc_notify] 0-glusterfsd-mgmt: failed to connect with remote-host: fs.glusterfs.service.consul (Success)
[2015-12-01 02:41:26.609831] I [glusterfsd-mgmt.c:1824:mgmt_rpc_notify] 0-glusterfsd-mgmt: Exhausted all volfile servers
[2015-12-01 02:41:26.610799] W [glusterfsd.c:1236:cleanup_and_exit] (-->/lib64/libgfrpc.so.0(rpc_clnt_notify+0xd0) [0x7f7aedc12d50] -->/usr/sbin/glusterfs(+0xf6be) [0x7f7aee31d6be] -->/usr/sbin/glusterfs(cleanup_and_exit+0x69) [0x7f7aee3176d9] ) 0-: received signum (1), shutting down
[2015-12-01 02:41:26.610842] I [fuse-bridge.c:5683:fini] 0-fuse: Unmounting '/mnt/container-volumes'.
[2015-12-01 02:41:26.613543] W [glusterfsd.c:1236:cleanup_and_exit] (-->/lib64/libpthread.so.0(+0x7df5) [0x7f7aeccacdf5] -->/usr/sbin/glusterfs(glusterfs_sigwaiter+0xe5) [0x7f7aee317855] -->/usr/sbin/glusterfs(cleanup_and_exit+0x69) [0x7f7aee3176d9] ) 0-: received signum (15), shutting down
```

In this case, running ansible a second time worked and glusterfs was up and running.

Looks like the consul service is not always ready or the glusterfs service is not yet registered/healthy by the time the `mount container volume` task runs. Perhaps we can add a wait_for condition on fs.glusterfs.service.consul

**Comment by justinclift** [comment]
> Perhaps we can add a wait_for condition on fs.glusterfs.service.consul

Definitely worth a shot. :smile:

**Comment by ryane** [comment]
fixed by #880
